lib/ABIAnalyses/DeadReturnValuesOfFunctionCall.cpp

```diff
@@ -0,0 +1,142 @@
+#include <cstdio>
+#include <iostream>
+#include <llvm/IR/Instructions.h>
+#include <llvm/Support/Casting.h>
+
+#include "revng/ABIAnalyses/DeadReturnValuesOfFunctionCall.h"
+#include "revng/MFP/MFP.h"
+#include "revng/Support/revng.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Instruction.h"
+
+namespace DeadReturnValuesOfFunctionCall {
+using namespace ABIAnalyses;
+using LatticeElement = MFI::LatticeElement;
+
+static State getMax(State Lh, State Rh) {
+  if (Lh == NoOrDead) {
+    return Rh;
+  } else if (Lh == Maybe) {
+    if (Rh == NoOrDead) {
+      return Lh;
+    } else {
+      return Rh;
+    }
+  } else {
+    return Lh;
+  }
+}
+
+LatticeElement MFI::combineValues(const LatticeElement &Lh,
+                                  const LatticeElement &Rh) const {
+
+  LatticeElement New = Lh;
+  for (const auto &KV : Rh) {
+    if (New.find(KV.first) != New.end()) {
+      New[KV.first] = getMax(New[KV.first], KV.second);
+    } else {
+      New[KV.first] = KV.second;
+    }
+  }
+  New.erase(InitialRegisterState);
+  return New;
+}
+
+bool MFI::isLessOrEqual(const LatticeElement &Lh,
+                        const LatticeElement &Rh) const {
+  if (Lh.size() > Rh.size())
+    return false;
+  for (auto &E : Lh) {
+    if (E.first == InitialRegisterState) {
+      continue;
+    }
+    if (Rh.count(E.first) == 0) {
+      return false;
+    }
+    auto RhState = Rh.lookup(E.first);
+    if (getMax(E.second, RhState) != RhState) {
+      return false;
+    }
+  }
+  return true;
+}
+
+LatticeElement MFI::applyTransferFunction(Label L,
+                                          const LatticeElement &E) const {
+  LatticeElement New = E;
+  for (auto &I : *L) {
+    switch (classifyInstruction(&I)) {
+    case TheCall: {
+      if (New.count(InitialRegisterState) != 0) {
+        // The first time we hit TheCall is actually the start of our graph
+        New.clear();
```

Comment:
This doesn't seem a good design. This should not be necessary.
Can't you start *right after* the call site (assuming it's a forward analysis)?

Reply:
Yes, but it's with the graphs that we are using (`GraphTraits<BasicBlock*>`) 

Some alternatives:
- checking if the current `BasicBlock` is the parent of the `CallSite` and skipping the first instructions
- instead of clearing, assigning a named constant like `InitialValue` or something similar

Reply:
Mmmmh, I see. But let's strive to avoid special casing at all costs.
Would splitting the basic block make things easier?
It'd be an harmless change. After all we are generating functions on the fly and then discarding them. If you need the IR in some form that does not disrupt what @antoniofrighetto is doing, let's go for it!

Reply:
yes having `Calls` on their own `BasicBlock` would be sufficient

Reply:
OK, let's go for it! `\o/`

Reply:
Now the analysis expects that each callsite is in its own basic block
So `InitialValue` is not needed anymore